File: frontend/src/services/resumeOptimizerService.js
import { apiClient } from "./apiClient";
import { API_URL } from "../config/api";

const BASE = "/api/resume-optimizer";

export const resumeOptimizerService = {
  getDashboard: () => apiClient.get(`${BASE}/dashboard`),
  listTemplates: () => apiClient.get(`${BASE}/templates`),
  listResumes: () => apiClient.get(BASE),
  getResume: (id) => apiClient.get(`${BASE}/${id}`),
  createResume: (data) => apiClient.post(BASE, data),
  updateResume: (id, data) => apiClient.put(`${BASE}/${id}`, data),
  deleteResume: (id) => apiClient.delete(`${BASE}/${id}`),
  duplicateResume: (id) => apiClient.post(`${BASE}/${id}/duplicate`),
  analyze: (id, data) => apiClient.post(`${BASE}/${id}/analyze`, data),
  improveSection: (id, data) => apiClient.post(`${BASE}/${id}/improve`, data),
  tailor: (id, data) => apiClient.post(`${BASE}/${id}/tailor`, data),
  applyChanges: (id, changes) => apiClient.post(`${BASE}/${id}/apply-changes`, { changes }),
  aiGenerate: (data) => apiClient.post(`${BASE}/ai-generate`, data),
  getJobMatches: (resumeId) => apiClient.get(`${BASE}/jobs/match${resumeId ? `?resumeId=${resumeId}` : ""}`),
  getCompanies: (targetRole) => apiClient.get(`${BASE}/companies${targetRole ? `?targetRole=${encodeURIComponent(targetRole)}` : ""}`),
  getSkillGaps: (targetRole) => apiClient.get(`${BASE}/skill-gaps${targetRole ? `?targetRole=${encodeURIComponent(targetRole)}` : ""}`),
  getInterviewContext: (id) => apiClient.get(`${BASE}/${id}/interview-context`),
  exportHtmlUrl: (id) => `${API_URL}${BASE}/${id}/export/html`,

  uploadResume: async (file, { title, templateId, targetRole } = {}) => {
    const token = localStorage.getItem("token");
    const form = new FormData();
    form.append("resume", file);
    if (title) form.append("title", title);
    if (templateId) form.append("templateId", templateId);
    if (targetRole) form.append("targetRole", targetRole);

    const res = await fetch(`${API_URL}${BASE}/upload`, {
      method: "POST",
      headers: token ? { Authorization: `Bearer ${token}` } : {},
      body: form,
    });
    const data = await res.json();
    if (!res.ok) throw new Error(data.message || "Upload failed");
    return data;
  },
};

export default resumeOptimizerService;

File: campus2career_mobile/lib/services/resume_optimizer_service.dart
import 'package:dio/dio.dart';
import '../models/resume_document.dart';
import 'api_helper.dart';

class ResumeOptimizerService {
  final ApiHelper _api;
  ResumeOptimizerService(this._api);

  static const _base = '/resume-optimizer';

  Future<CareerDashboard> getDashboard() async {
    final data = await _api.get<Map<String, dynamic>>('$_base/dashboard');
    return CareerDashboard.fromJson(data['dashboard'] is Map ? Map<String, dynamic>.from(data['dashboard']) : {});
  }

  Future<List<ResumeDocument>> listResumes() async {
    final data = await _api.get<Map<String, dynamic>>('$_base');
    final list = data['resumes'];
    if (list is! List) return [];
    return list.whereType<Map>().map((e) => ResumeDocument.fromJson(Map<String, dynamic>.from(e))).toList();
  }

  Future<ResumeDocument> getResume(String id) async {
    final data = await _api.get<Map<String, dynamic>>('$_base/$id');
    return ResumeDocument.fromJson(Map<String, dynamic>.from(data['resume'] as Map));
  }

  Future<ResumeDocument> createResume({String? title, bool fromProfile = true, String? targetRole}) async {
    final data = await _api.post<Map<String, dynamic>>('$_base', body: {
      'title': title ?? 'My Resume',
      'fromProfile': fromProfile,
      if (targetRole != null) 'targetRole': targetRole,
    });
    return ResumeDocument.fromJson(Map<String, dynamic>.from(data['resume'] as Map));
  }

  Future<ResumeDocument> updateResume(String id, {String? title, Map<String, dynamic>? content, String? targetRole}) async {
    final data = await _api.put<Map<String, dynamic>>('$_base/$id', body: {
      if (title != null) 'title': title,
      if (content != null) 'content': content,
      if (targetRole != null) 'targetRole': targetRole,
    });
    return ResumeDocument.fromJson(Map<String, dynamic>.from(data['resume'] as Map));
  }

  Future<ResumeDocument> uploadResume({required String filePath, String? filename, String? title}) async {
    final data = await _api.postMultipart<Map<String, dynamic>>(
      '$_base/upload',
      fileField: 'resume',
      filePath: filePath,
      filename: filename,
      fields: {if (title != null) 'title': title},
    );
    return ResumeDocument.fromJson(Map<String, dynamic>.from(data['resume'] as Map));
  }

  Future<Map<String, dynamic>> analyze(String id, {String? jobDescription}) async {
    final data = await _api.post<Map<String, dynamic>>('$_base/$id/analyze', body: {
      if (jobDescription != null) 'jobDescription': jobDescription,
    });
    return data['analysis'] is Map ? Map<String, dynamic>.from(data['analysis']) : {};
  }

  Future<List<Map<String, dynamic>>> getJobMatches({String? resumeId}) async {
    final data = await _api.get<Map<String, dynamic>>('$_base/jobs/match', query: {
      if (resumeId != null) 'resumeId': resumeId,
    });
    final list = data['matches'] ?? data['jobs'];
    if (list is! List) return [];
    return list.whereType<Map>().map((e) => Map<String, dynamic>.from(e)).toList();
  }

  Future<List<Map<String, dynamic>>> getCompanies({String? targetRole}) async {
    final data = await _api.get<Map<String, dynamic>>('$_base/companies', query: {
      if (targetRole != null) 'targetRole': targetRole,
    });
    final list = data['companies'];
    if (list is! List) return [];
    return list.whereType<Map>().map((e) => Map<String, dynamic>.from(e)).toList();
  }

  Future<Map<String, dynamic>> getSkillGaps({String? targetRole}) async {
    final data = await _api.get<Map<String, dynamic>>('$_base/skill-gaps', query: {
      if (targetRole != null) 'targetRole': targetRole,
    });
    return data['gaps'] is Map ? Map<String, dynamic>.from(data['gaps']) : {};
  }

  Future<Map<String, dynamic>> getInterviewContext(String id) async {
    final data = await _api.get<Map<String, dynamic>>('$_base/$id/interview-context');
    return data['context'] is Map ? Map<String, dynamic>.from(data['context']) : {};
  }
}
